### ISSUE TITLE
fastai v1: create_cnn hangs on Window10

### ISSUE THREAD

**Issue by parthopdas** [fastai/fastai#1738]
**Describe the bug**

fastai v1: create_cnn hangs on Window10. [Here](https://forums.fast.ai/t/lesson-1-notebook-stuck-in-create-cnn/37486/23) is the discussion on. Filing this issue so it is on the radar.

**Provide your installation details**

```text
=== Software ===
python        : 3.7.2
fastai        : 1.0.46
fastprogress  : 0.1.20
torch         : 1.0.1
nvidia driver : 417.35
torch cuda    : 10.0 / is available
torch cudnn   : 7401 / is enabled

=== Hardware ===
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 4096MB | GeForce GTX 1050 Ti with Max-Q Design

=== Environment ===
platform      : Windows-10-10.0.17763-SP0
conda env     : my-fastai2019
python        : C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019\python.exe
sys.path      :
C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019\python37.zip
C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019\DLLs
C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019\lib
C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019
C:\Users\partho\AppData\Local\conda\conda\envs\my-fastai2019\lib\site-packages

Sun Mar 03 23:57:11 2019
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 417.35       Driver Version: 417.35       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 105... WDDM  | 00000000:01:00.0  On |                  N/A |
| N/A   50C    P3    N/A /  N/A |   2701MiB /  4096MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1360    C+G   ...8.0_x64__8wekyb3d8bbwe\WinStore.App.exe N/A      |
|    0      1588    C+G   Insufficient Permissions                   N/A      |
|    0      2496    C+G   ...t_cw5n1h2txyewy\ShellExperienceHost.exe N/A      |
|    0      3984    C+G   ...mmersiveControlPanel\SystemSettings.exe N/A      |
|    0      4092    C+G   ...dows.Cortana_cw5n1h2txyewy\SearchUI.exe N/A      |
|    0      6448    C+G   ...x86__k1h2ywk1493x8\Lenovo.Discovery.exe N/A      |
|    0      8184    C+G   C:\Windows\explorer.exe                    N/A      |
|    0     10744    C+G   ...40.70.0_x64__kzf8qxf38zg5c\SkypeApp.exe N/A      |
|    0     11132    C+G   ...0453.0_x64__8wekyb3d8bbwe\YourPhone.exe N/A      |
|    0     11500    C+G   ...osoft.LockApp_cw5n1h2txyewy\LockApp.exe N/A      |
|    0     12552    C+G   ...partho\scoop\apps\hyper\2.1.2\Hyper.exe N/A      |
|    0     12820    C+G   ...hell.Experiences.TextInput.InputApp.exe N/A      |
|    0     13300    C+G   ...6)\Google\Chrome\Application\chrome.exe N/A      |
|    0     14292    C+G   ...oftEdge_8wekyb3d8bbwe\MicrosoftEdge.exe N/A      |
|    0     14968    C+G   ...x64__8wekyb3d8bbwe\Microsoft.Photos.exe N/A      |
|    0     17084    C+G   ...artho\scoop\apps\vscode\1.31.1\Code.exe N/A      |
+-----------------------------------------------------------------------------+

```

**To Reproduce**
1. conda create --name my-fastai2019 python=3.7.2
2. conda activate my-fastai2019
3. conda install -c pytorch -c fastai -c conda-forge fastai pytorch torchvision cuda100
4. conda install jupyter
5. run lesson1-pets.ipynb

The notebook hangs at the create_cnn step.

**Expected behavior**
Notebook to complete successfully.

**Screenshots**
none

**Additional context**
However if we uninstall pytorch 1.0.1 and install pytorch 1.0.0, the above issue does not happen. I can confirm this. In addition [others](https://forums.fast.ai/t/lesson-1-notebook-stuck-in-create-cnn/37486/23) have confirmed the same on the forum.

**Comment by sgugger** [comment]
If the freeze happens with pytorch 1.0.1 and not 1.0.0 it seems like a bug in pytorch, no? Could you maybe try to execute the code in `create_cnn` line by line to figure out which instruction is causing the freeze?

**Comment by parthopdas** [comment]
@sgugger possibly. Since the issue is exposed though fastai, I have filed it here.

I'll debug into create_cnn and report back with more info.

**Comment by sgugger** [comment]
Did you have any new on this?

**Comment by parthopdas** [comment]
My apologies for the delay.

I have the repro running under the debugger. Code paths are unfamiliar so I need some help.
Perhaps a screen share over a call with get this done waaay faster.

Also I am not proficient in debugging python - so please let me know if I can collect additional debugger info for you.

Here is what is leading to the hang:

```python
# start
learn = create_cnn(data, models.resnet34, metrics=error_rate)
# ->
nf = num_features_model(nn.Sequential(*body.children())) * 2
# ->
        try: return model_sizes(m, size=(sz,sz))[-1][1] # sz = 64
# ->
        x = dummy_eval(m, size)
# ->
    return m.eval()(dummy_batch(m, size))
# ->
    return one_param(m).new(1, ch_in, *size).requires_grad_(False).uniform_(-1.,1.) # ch_in = 3
# ->
    return next(m.parameters())
# ->
        self.stored = self.hook_func(module, input, output)
```
> At this point variables on the stack are as follows

| Name | Value | Type |
|-|-|-|
| > | input | <generator object Hook.hook_fn.<locals>.<genexpr> at 0x0000027D3AF5EB10> | generator |
| > | module | Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False) | Conv2d |
| > | output | tensor([[[[ 2.7787e-01,  7.0241e-01,  8.2423e-02,  ..., -3.2922e-01,             1| Tensor |
| > | self | <fastai.callbacks.hooks.Hook object at 0x0000027D3AF5C390> | Hook |

```python
# resuming the above call 
        self.stored = self.hook_func(module, input, output)
# ->
        self.stored = self.hook_func(module, input, output)
# ->
        self.stored = self.hook_func(module, input, output)
# ->
        self.stored = self.hook_func(module, input, output)
```

So the 5th call to self.hook_func hangs.

**Comment by sgugger** [comment]
Thanks for your input. I managed to reproduce the bug and create a minimal reproducible example in pure PyTorch. There is an [issue filed](https://github.com/pytorch/pytorch/issues/17840) that you can follow.

In the meantime, a workaround is to set 
```
torch.set_num_threads(1)
```

**Comment by sgugger** [comment]
We removed the line that was setting threads from fastai and it removes the bug from Windows. Will make a release soon so that it's available for all users.

**Comment by SSDL2018** [comment]
so with the new release we don't need to use 

> torch.set_num_threads(1)

 anymore right?

**Comment by sgugger** [comment]
Yup, this isn't necessary anymore.
